text/0000-proc-macro-include.md

```diff
@@ -0,0 +1,205 @@
+- Improve proc macro sandboxability and cacheability, by offering a way to implement this class of
+  file-reading macros without using OS APIs directly.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## For users of proc macros
+
+Nothing changes! You'll just see nicer errors and fewer rebuilds
+from procedural macros which read external files.
+
+## For writers of proc macros
+
+Three new functions are provided in the `proc_macro` interface crate:
+
+```rust
+/// Read the contents of a file as a `TokenStream` and add it to build dependency graph.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// May fail for a number of reasons, for example, if the string contains unbalanced delimiters
+/// or characters not existing in the language.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// Source spans are constructed for the read file. If you use the spans of this token stream,
+/// any resulting errors will correctly point at the tokens in the read file.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include<P: AsRef<str>>(path: P) -> Result<TokenStream, std::io::Error>;
+
+/// Read the contents of a file as a string literal and add it to build dependency graph.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include_str<P: AsRef<str>>(path: P) -> Result<Literal, std::io::Error>;
+
+/// Read the contents of a file as raw bytes and add it to build dependency graph.
+///
+/// The build system executing the compiler will know that the file was accessed during compilation,
+/// and will be able to rerun the build when the contents of the file changes.
+///
+/// If the file fails to be read, this is not automatically a fatal error. The proc macro may
+/// gracefully handle the missing file, or emit a compile error noting the missing dependency.
+///
+/// NOTE: some errors may cause panics instead of returning `io::Error`.
+/// We reserve the right to change these errors into `io::Error`s later.
+fn include_bytes<P: AsRef<str>>(path: P) -> Result<Vec<u8>, std::io::Error>;
```

Comment:
Would it make sense for `include_bytes` to return `Literal` as well, or would that not be possible?

Reply:
I think it should work because `Literal` can be a byte string.

Reply:
Actually, yeah, I overlooked that possibility.

The main limitation is that the only current interface for getting the contents *out* of a `Literal` is to `ToString` it. `syn` does have a `.value()` for `LitByteStr` as well as `LitStr`, though, so I guess it's workable.

It's probably not *good* to short term require debug escaping a binary file to reparse the byte string literal if a proc macro is going to post process the file... but if it's just including the literal, it can put the `Literal` in the token stream, and we can offer ways to extract (byte) string literals without printing the string literal in the future.

Reply:
The one limitation which needs to be solved is *how do spans work*. Do we just say that the byte string literal contains the raw bytes of the file (even though that would be illegal in a normal byte string, and invalid UTF-8), maybe as a new "kind" of byte string, so span offsets are mapped directly with the source file? Or are there multiple span positions (representing a `\xNN` in the byte string) which map to a single byte in the source file?

Reply:
Hmm, what bytes are not allowed in byte string literals? Does the literal itself have to be valid UTF-8?

Reply:
A Rust source file must be valid UTF-8. Thus, the contents of a byte string literal in the source must be valid UTF-8.

Bytes that are not < 0x80 thus must be escaped to appear in a byte string literal.

Reply:
And then another question that's worth making explicit: what does it even mean for rustc to report a span into a binary file?

I think binary includes are better served by a different API that lets rustc point into generated code, rather than trying to point into an opaque binary file.